### ISSUE TITLE
Update PHP-Scoper to v0.16.2

### ISSUE THREAD

**Issue by swissspidy** [swissspidy/web-stories-wp#10407]
## Task Description

The current version of PHP-Scoper is 0.16.2.

In [version 0.16.0](https://github.com/humbug/php-scoper/releases/tag/0.16.0), there were lots of improvements to the config, making it a lot simpler for us to configure PHP-Scoper in a WP plugin context.

It means we can get rid of all the manual `str_replace` calls in the scoper config thanks to `expose-global-{constants, classes, functions}` and `expose-{namespaces, constants, classes, functions}`

We could even automate this by setting the config based on the WP stubs, see https://github.com/humbug/php-scoper/issues/303#issuecomment-1021388471

Though we could split that up in its own ticket.

### Implementation Brief

* Install latest version of PHP-Scoper via Composer
* Remove all the `str_replace` calls in the PHP-Scoper config
* Install [`sniccwp/php-scoper-wordpress-excludes`](https://packagist.org/packages/sniccwp/php-scoper-wordpress-excludes) as a dev dependency and configure the `exclude-classes`, `exclude-functions`, and `exclude-constants` options accordingly

**Comment by spacedmonkey** [comment]
So is the recommendation here to use these packages? 

- [sniccowp/php-scoper-wordpress-excludes](https://github.com/sniccowp/php-scoper-wordpress-excludes/tree/5.8.2)
- [sniccowp/php-scoper-wp-cli-excludes](https://github.com/sniccowp/php-scoper-wp-cli-excludes/tree/2.5.0)

Then removing this ? 

https://github.com/GoogleForCreators/web-stories-wp/blob/d16f28fb45d808bf5abef2a5ae2c69feb5a63f66/scoper.inc.php#L176-L258

**Comment by swissspidy** [comment]
First of all we should remove all our `str_replace()` calls and use the new config options `exclude-functions`, `exclude-constants`, et al.

That should be trivial. Just an array of functions and constants and class names.

Afterwards, most likely in its own ticket, we can look into automating it somehow. It needs more thinking on how best to solve that part.

I would like to avoid using these packages you mentioned. They are not really well maintained and not up-to-date, just proof-of-concepts. However, the underlying https://github.com/sniccowp/php-scoper-excludes utility shows how it could be done.

**Comment by calvinalkan** [comment]
@swissspidy Hey there, I'm the creator of these packages.

We'll maintain them for a good time to come, php-scoper is very vital to a big framework we are going to release for WordPress.

I'd like to automate the release of the stubs, but I don't know quite now what the best approach would be since it requires php-stubs/wordpress-stubs. 
I thought about a Github Action with a corn schedule. If you have any suggestions I'm all ears. I think this is something that desperately needs solving for the WP plugin ecosystem.

**Comment by swissspidy** [comment]
@calvinalkan hey there, thanks for chiming in! Since it requires `php-stubs/wordpress-stubs` and everyone might have a different version of that package, I was wondering whether the exclusion rules generation could be done as part of a composer plugin or something. Perhaps it could even be done on-the-fly in the PHP-Scoper config file itself. After all the scoping only happens on `composer install` anyway (at least in our case), so that's not gonna be a huge issue.

A GiHub actions sounds interesting too, but then I'd probably set one up for ourselves in this repo, so that we don't have to require an extra composer package for it.

**Comment by calvinalkan** [comment]
Okay so there two use cases.

1. Directly require --dev [sniccowp/php-scoper-wordpress-excludes](https://github.com/sniccowp/php-scoper-wordpress-excludes/tree/5.8.2)

with whatever version tag you need for your project. IMO this should always be the latest stable tag of WordPress itself that's why Id like to automate this to the maximum degree possible so that your plugin can require it as

```json

"require-dev": {
  "sniccowp/php-scoper-wordpress-excludes" : "^5.0|^6.0"
}

```

And then in your scoper.inc.php you just require:

```php

'exclude_classes' => array_merge(
  require $path_to_vendor.'/sniccowp/php-scoper-excludes/generated/wordpress-classes.php',
  ... other classes
)

```
Now, in your build step, you always exclude against the latest version of WP.


2. If you need to generate custom excludes for something different (maybe a third-party plugin that you integrate with) than the wp, wp-cli, and woocommerce packages that already exist, you should directly require the [cli-app](https://github.com/sniccowp/php-scoper-excludes) as a dev-dependency (or composer plugin) and generate your exclusion list once and put them into VCS. You could also build the exclusion list in CI but I don't know if that adds any benefits.

**Comment by calvinalkan** [comment]
See [here](https://github.com/sniccowp/php-scoper-wordpress-excludes/blob/5.8.2/generate-excludes.inc.php) how the exclusion lists for WordPress itself are generated. 

But yeah, for the core/wp-cli excludes I don't see why you wouldn't just require the list of generated files as a dev dependency. The package is basically just 5 files.

**Comment by swissspidy** [comment]
If you're gonna automate maintenance of your `sniccowp/php-scoper-wordpress-excludes` package like you mentioned, that'd be great. Otherwise I was merely wondering whether we couldn't just inline all of this work in this repository so that we don't need this extra dependency.

**Comment by calvinalkan** [comment]
My initial idea was: 

- A GitHub action that runs on a daily cron
- fetch latest tag of php-stubs/wordpress-stubs
- compare with a current_version.txt in the repo
- update if needed
- tag new release
- update latest_version.txt 

Do you have any other idea?

**Comment by swissspidy** [comment]
You wouldn't need `current_version.txt` if you just look at all the existing Git tags on your end.

**Comment by calvinalkan** [comment]
@swissspidy [Implemented now.](
https://github.com/sniccowp/php-scoper-wordpress-excludes/blob/master/.github/workflows/update.yml) We run a daily check against php-stubs/wordpress stubs and tag a new release automatically. Available at [packagist](https://packagist.org/packages/sniccwp/php-scoper-wordpress-excludes) now aswell.

We changed from .php to .json because with .php git would go mad on the numerically indexed array. It would show 6000 additions and deletions basically. With JSON it's much clearer now what got updated and what got removed.
